elf.c

```diff
@@ -78,11 +80,19 @@ static unsigned long read_elf64(int fd)
 	ehdr.e_shentsize	= file16_to_cpu(ehdr64.e_shentsize);
 	ehdr.e_shnum		= file16_to_cpu(ehdr64.e_shnum);
 
-	return(ehdr.e_shoff + (ehdr.e_shentsize * ehdr.e_shnum));
+	offset = ehdr.e_shoff + (ehdr.e_shentsize * (ehdr.e_shnum - 1));
+	ret = pread(fd, &shdr64, sizeof(shdr64), offset);
+	if (ret < 0 || (size_t)ret != sizeof(ehdr)) {
+		fprintf(stderr, "Read of ELF section header from %s failed: %s\n",
+			fname, strerror(errno));
+		exit(10);
+	}
+
+	return(file64_to_cpu(shdr64.sh_offset) + file64_to_cpu(shdr64.sh_size));
```

Comment:
Executable ends where the data of its last section ends, which is the sum of its offset and its length.